KEYCLOAK-11472 Pagination support for clients

saibot94 opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

This MR addresses issues:
- https://issues.jboss.org/browse/KEYCLOAK-11472 (related to the one below)
- https://issues.jboss.org/browse/KEYCLOAK-8275 

## Details

**TL;DR:**
There are a few performance concerns that we will hit once we go production with our service at CERN (more than 15k clients on one realm).

More details can be read on my [keycloak-dev mailing list mail](https://lists.jboss.org/pipermail/keycloak-dev/2019-September/012487.html) about the performance issues.

## Tests
I ran the unit + integration tests locally, got:

```
[ERROR] Failures: 
[ERROR]   UserStorageConsentTest.testLogin:177 URL expected to begin with: https://localhost:8543/auth/realms/demo/protocol/openid-connect/auth ; actual URL: http://localhost:8280/product-portal/
[ERROR]   ClientRedirectTest.testRedirectToDisabledClientRedirectURI:119 type
Expected: is "LOGOUT_ERROR"
     but: was "LOGOUT"
[ERROR] Errors: 
[ERROR]   ResourceServerManagementTest.testCreateAndDeleteResourceServer:66 » InternalServerError
[ERROR]   RealmTest.renameRealm:130 » InternalServerError HTTP 500 Internal Server Error
[ERROR]   JsonFileImport198MigrationTest.migration1_9_8Test:68->AbstractMigrationTest.testMigrationTo2_0_0:133->AbstractMigrationTest.testAuthorizationServices:397 » InternalServerError
[INFO] 
[ERROR] Tests run: 3193, Failures: 2, Errors: 3, Skipped: 196
```

Not sure if my changes.

saibot94 committed:
Client search endpoint, pagination and performance improvements

stianst reviewed (changes requested):
See thread on dev ML

stianst commented:
Changes done to logout is a breaking change and as I suggested in the ML we should discuss that separately with a separate issue/PR. Ideal solution is to introduce a new logout endpoint that is compliant to OIDC session management, deprecating the old logout endpoint, updating all adapters to use the new endpoint. The finally in a while we can remove the old logout endpoint, but those that don't want to have the performance overhead of the old endpoint with large number of clients should obviously avoid it. We can also have a feature toggle to allow disabling the old logout endpoint.

stianst commented:
Updated the JIRA description for KEYCLOAK-11472. This PR should be limited to cover that issue and not to everything in one go.

saibot94 commented:
I'll uncomment the logout part, just leaving the clients backend / frontend changes. 

Thanks for the review.

saibot94 commented:
@stianst thanks for the review.

I uncommented the logout thing, so that we track it in another issue.

If you could have another look, the `ClientsResource` search is implemented in a similar way to the other paginated endpoints (Users, etc). It still has backwards compatibility, in the sense that if you omit `first` and `max` it behaves like before.

The more major changes are in the UI, I'd suggest you build the server and go around through the various places needing a client list. I tested them and they seem to work like previously, but an extra pair of eyes is needed.

If there should be any unit / integration tests to write, please let me know where to get started with.

saibot94 commented:
Hey @stianst, I was checking the master builds but it seems the same `crossdc-server` tests are still failing on master (I rebased on Friday): https://travis-ci.org/keycloak/keycloak/jobs/588125883 

Please let me know when you manage to review the changes in this PR.

stianst commented:
We're having some issues with the cross-dc tests, so it's most likely not related to this PR. I will try to get someone to review this as soon as we can, but we do have a bit of a list of PRs to get through at the moment I'm afraid.

stianst commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmAdapter.java:
```diff
@@ -680,8 +680,7 @@ public void removeDefaultRoles(String... defaultRoles) {
     public List<ClientModel> getClients() {
-        return cacheSession.getClients(this);
-
+        return cacheSession.getClients(this, 0, cacheSession.getClientsCount(this).intValue());
```
Realm should have an overloaded method getClients(this), that just calls getClients(this, null, null). This approach will result in extra query to get the client counts first, to then list all clients.

stianst commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java:
```diff
@@ -537,40 +537,8 @@ static String getRoleByNameCacheKey(String container, String name) {
     }
```
Just checking here - we are removing any caching of client lists, and only having caching of individual clients?

saibot94 replied:
Yes, that was the idea. I thought we shouldn't cache the pages when it comes to getting clients, but if you think this should change, let me know!

stianst replied:
I was mostly just checking as we shouldn't cache clients at all. Client lists are only used in admin console, and while being used in admin console often changed. So it doesn't make any sense to cache client lists.

stianst commented on model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:
```diff
@@ -21,30 +21,16 @@
-import org.keycloak.models.jpa.entities.GroupEntity;
-import org.keycloak.models.jpa.entities.RealmEntity;
-import org.keycloak.models.jpa.entities.RoleEntity;
+import org.keycloak.models.*;
```
Please don't use import ...*.

Also, make sure there are no changes that are not needed in your PR as it makes it much harder to review.

stianst commented on model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:
```diff
@@ -304,7 +289,7 @@ public boolean removeRole(RealmModel realm, RoleModel role) {
         String compositeRoleTable = JpaUtils.getTableNameForNativeQuery("COMPOSITE_ROLE", em);
         em.createNativeQuery("delete from " + compositeRoleTable + " where CHILD_ROLE = :role").setParameter("role", roleEntity).executeUpdate();
-        realm.getClients().forEach(c -> c.deleteScopeMapping(role));
+        realm.getClients(0, realm.getClientsCount().intValue()).forEach(c -> c.deleteScopeMapping(role));
```
Same comment as earlier. We need a realm.getClients() that can just call realm.getClients(null,null) without pagination instead of fetching count first.

stianst commented on model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:
```diff
@@ -547,11 +540,17 @@ public ClientModel getCreatedClient() {
         query.setParameter("realm", realm.getId());
         List<String> clients = query.getResultList();
-        if (clients.isEmpty()) return Collections.EMPTY_LIST;
+        if (clients.isEmpty()) return Collections.emptyList();
```
Please don't make unnecessary changes (Collections.EMPTY_LIST is perfectly valid)

saibot94 replied:
Sure, I changed it back. I'm using an IDE for the development and it was giving me an `Unchecked assignment` warning, since `Collections.EMPTY_LIST` returns the non-generic version.

stianst commented on model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:
```diff
@@ -582,6 +581,22 @@ public ClientModel getClientByClientId(String clientId, RealmModel realm) {
+        query.setParameter("clientId", clientId);
+        query.setParameter("realm", realm.getId());
+        List<String> results = query.getResultList();
+        if (results.isEmpty()) return null;
```
This should return empty list rather than null

stianst commented on model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java:
```diff
@@ -779,10 +784,14 @@ public void removeDefaultGroup(GroupModel group) {
     @Override
     public List<ClientModel> getClients() {
-        return session.realms().getClients(this);
+        return session.realms().getClients(this, 0, session.realms().getClientsCount(this).intValue());
```
getClients(this, null, null)

stianst commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java:
```diff
@@ -104,6 +104,7 @@ public Response logout(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_PARAM) String
                            @QueryParam("initiating_idp") String initiatingIdp) {
         String redirect = postLogoutRedirectUri != null ? postLogoutRedirectUri : redirectUri;
 
+        // TODO evaluate importance of removing this: major performance boost observed
```
Remove this TODO as it could result in someone removing this by mistake, and doing so would be a breaking change. We are planning on introducing support for RP Initiated logout (https://issues.jboss.org/browse/KEYCLOAK-11603) which won't suffer from the design issue that this logout endpoint does. We will deprecate and eventually remove the legacy KC endpoint.

saibot94 replied:
Removed, thanks!

stianst commented on services/src/main/java/org/keycloak/services/resources/admin/ClientInitialAccessResource.java:
```diff
@@ -28,13 +28,7 @@
-import javax.ws.rs.Path;
-import javax.ws.rs.PathParam;
-import javax.ws.rs.Produces;
+import javax.ws.rs.*;
```
Please don't change imports to ...*

stianst commented on services/src/main/java/org/keycloak/services/resources/admin/ClientsResource.java:
```diff
@@ -16,21 +16,13 @@
-import org.keycloak.models.ModelDuplicateException;
-import org.keycloak.models.RealmModel;
-import org.keycloak.models.UserModel;
+import org.keycloak.models.*;
```
Please don't change imports to ...*

stianst commented on themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties:
```diff
@@ -1214,10 +1215,12 @@ group-membership.tooltip=Groups user is a member of. Select a listed group and c
 membership.available-groups.tooltip=Groups a user can join. Select a group and click the join button.
 table-of-realm-users=Table of Realm Users
 view-all-users=View all users
+view-all-clients=View all clients
```
The clients page should just display the first list of clients, not requiring clicking view all clients.

stianst reviewed (changes requested):
Would like the clients list to just list the first page of clients, not requiring clicking on view all clients.

Found at least one mapper that isn't quite working. Protocol mappper user client role will just display first page of clients it seems, so I wasn't able to select from all clients here. Guess this should be changed to have the search for client like other mappers.

stianst commented:
@saibot94 can you rebase please? Also, did you try to run the admin console tests on this one?

vmuzikar reviewed (changes requested):
@saibot94 Thank you for your contribution! It's a really nice improvement. ;)

I've done some quick preliminary review and I've found some issues with the UI tests.

The following tests are failing:
* `ClientClientScopesTest.testEvaluateClientScopes`
* `ClientMappersOIDCTest.testHardcodedRole`
* `ClientsTest.clientsCRUD`

I believe those are just false positives caused by the new "View all clients" button.

Also, do you think you could add some UI tests for the pagination, similar to [other clients UI tests](https://github.com/keycloak/keycloak/tree/master/testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients)? [Here](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/HOW-TO-RUN.md#admin-console-ui-tests) you can find some more details about the UI tests and how to run them with different browsers.

@stianst We'll also need to update screenshots in the docs. But I believe this is out of scope here, right?

saibot94 committed:
Address comments on PR

saibot94 committed:
Removed 'View all clients' buton from the UI

saibot94 commented:
@vmuzikar Thank you for the comments!

I wanted to add some tests but I'm blocked at the running part.

I run:

```
mvn -f testsuite/integration-arquillian/tests/other/console/pom.xml \
    clean test \
    -Dbrowser=firefox \
    -Dfirefox_binary=/usr/bin/firefox
```
and I get an exception thrown at the start of the test run:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (default-test) on project integration-arquillian-tests-console: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/cristi/work/keycloak/testsuite/integration-arquillian/tests/other/console/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /home/cristi/work/keycloak/testsuite/integration-arquillian/tests/other/console && /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -Djava.awt.headless=true -Xms512m -Xmx2048m -XX:MetaspaceSize=96m -XX:MaxMetaspaceSize=256m -jar /home/cristi/work/keycloak/testsuite/integration-arquillian/tests/other/console/target/surefire/surefirebooter7004083962389781767.jar /home/cristi/work/keycloak/testsuite/integration-arquillian/tests/other/console/target/surefire 2019-10-22T09-28-56_557-jvmRun1 surefire1120209712289794202tmp surefire_07813822963043121705tmp
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 1
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /home/cristi/work/keycloak/testsuite/integration-arquillian/tests/other/console && /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -Djava.awt.headless=true -Xms512m -Xmx2048m -XX:MetaspaceSize=96m -XX:MaxMetaspaceSize=256m -jar /home/cristi/work/keycloak/testsuite/integration-arquillian/tests/other/console/target/surefire/surefirebooter7004083962389781767.jar /home/cristi/work/keycloak/testsuite/integration-arquillian/tests/other/console/target/surefire 2019-10-22T09-28-56_557-jvmRun1 surefire1120209712289794202tmp surefire_07813822963043121705tmp
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 1
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:671)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:533)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:278)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:244)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1194)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1022)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:868)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:956)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:192)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)

```

Any help would be appreciated. Meanwhile, I removed the "View all clients" button and it should work as expected. If you can, please run those failing UI tests again.

saibot94 commented:
@stianst Thanks for the review! 

I addressed most of the comments, except the one with the caching of the client lists.

Removed the `import *` statements, rebased on master, the clients page has no more `View all clients` button and there's an implementation of `getClients()` which calls `getClients(realm, null, null)`, as you suggested.

saibot94 committed:
Update remaining client ID search components

saibot94 committed:
Fix remaining failing UI tests.

saibot94 committed:
Add clients page navigation test

saibot94 commented:
@vmuzikar I've implemented a comprehensive test that moves around the client pages and fixed the remaining failing ones. 

You can have another look when you have some time!

vmuzikar commented on services/src/main/java/org/keycloak/storage/openshift/OpenshiftClientStorageProvider.java:
```diff
@@ -73,6 +75,12 @@ public ClientModel getClientByClientId(String clientId, RealmModel realm) {
+    @Override
+    public List<ClientModel> searchClientsByClientId(String clientId, Integer firstResult, Integer maxResults, RealmModel realm) {
+        // TODO not sure about this, but I don't see this implementation using the search now
+        return Collections.singletonList(getClientByClientId(clientId, realm));
```
@saibot94 @stianst I'm not sure about this. Does this mean, if we want to select a Client e.g. in some Client Role selector and this Client is an Openshift Client, we need to exactly match the Client ID to select it?

stianst replied:
I'm okay with leaving it not supporting search for now. The OpenshiftClientStorageProvider is not supported at the moment and most likely needs a bit of work regardless when we pick up the effort to be able to replace IdP in OpenShift with Keycloak.

vmuzikar commented on testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/mappers/CreateClientMappersForm.java:
```diff
@@ -117,14 +118,21 @@ public void selectRealmRole(String roleName) {
+            driver.findElement(By.cssSelector("#s2id_clients .select2-choice")).click();
+            String subContainerClass = "#select2-drop:not([style*='display: none'])";
+            driver.findElement(By.cssSelector( subContainerClass + " .select2-input")).sendKeys(clientName);
+            driver.findElements(By.cssSelector(subContainerClass + " .select2-results li.select2-result-selectable")).get(0).click();
```
Could you perhaps try to use similar approach to [this](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/authorization/policy/ClientPolicyForm.java#L89)? The point is not to copy-paste code between Page Objects and also not to put so many ugly selectors directly into the methods.

vmuzikar replied:
@saibot94 Please, let me know if you're working on this and if I can help you with this somehow. It's actually not really that blocking and I'm willing to approve this as is. It's just something nice to have as it's definitely a reusable peace of code as the Client selector is now used at many places throughout the Admin Console – so it'd make sense to make it a Page Fragment as I mentioned.

saibot94 replied:
Hey, you can check out the `ClientSelectModal` class I've added. It hides away the ugly CSS selectors in `ClientScopesEvaluateForm` and `CreateClientMappersForm`.

I think you wanted something like this, the tests run just fine.

vmuzikar commented on testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/clientscopes/ClientScopesEvaluateForm.java:
```diff
@@ -103,8 +103,9 @@ public void setAssignedClientScopes(Collection<String> scopes) {
-        driver.findElement(By.className("select2-result-label")).click();
+        String subContainerClass = "#select2-drop:not([style*='display: none'])";
+        driver.findElement(By.cssSelector( subContainerClass + " .select2-input")).sendKeys(username);
+        driver.findElements(By.cssSelector(subContainerClass + " .select2-results li.select2-result-selectable")).get(0).click();
```
Could you perhaps try to use similar approach to [this](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/authorization/policy/ClientPolicyForm.java#L89)? The point is not to copy-paste code between Page Objects and also not to put so many ugly selectors directly into the methods.

saibot94 replied:
Hey, you can check out the `ClientSelectModal` class I've added. It hides away the ugly CSS selectors in `ClientScopesEvaluateForm` and `CreateClientMappersForm`.

I think you wanted something like this, the tests run just fine.

vmuzikar reviewed (changes requested):
@saibot94 Thank you for the update! I'm sorry for the delay.

I've done some extensive testing, both automated and manual, and I've found a few (hopefully last :)) issues.

saibot94 committed:
Fix client role selector in modal

saibot94 commented on themes/src/main/resources/theme/base/admin/resources/partials/modal/role-selector.html:
```diff
@@ -19,13 +19,15 @@ <h4 class="modal-title">{{:: 'role-selector' | translate}}</h4>
+                <kc-tooltip>{{:: 'client-roles.tooltip' | translate}}</kc-tooltip>        
             </label>
+            <div class="col-md-6">
+                <input type="hidden" ui-select2="clientsUiSelect" id="clients" data-ng-model="selectedClient" data-ng-change="changeClient(selectedClient);" data-placeholder="{{:: 'authz-select-client' | translate}}...">
```
Fixed, thanks!

vmuzikar replied:
Thank you, looks good!

saibot94 committed:
Fix default role selector

saibot94 commented on themes/src/main/resources/theme/base/admin/resources/partials/realm-default-roles.html:
```diff
@@ -44,15 +44,13 @@ <h1>{{:: 'roles' | translate}}</h1>
-                    <div class="col-md-4"><span class="text-muted">{{:: 'select-client-to-view-roles' | translate}}</span></div>
+           <label class="col-md-2 control-label" for="clients">{{:: 'client-roles' | translate}}</label>
+                <div class="col-md-6">
+                    <input type="hidden" ui-select2="clientsUiSelect" id="clients" data-ng-model="selectedClient" data-ng-change="changeClient(selectedClient);" data-placeholder="{{:: 'authz-select-client' | translate}}...">
```
@vmuzikar thanks for spotting it!

It should be working according to expectations now, see the attached photo.

![image](https://user-images.githubusercontent.com/8788869/68084070-07400700-fe31-11e9-9f39-0914a0522d88.png)

vmuzikar replied:
Thank you, looks good!

saibot94 committed:
Add ClientSelectModal to hide ugly CSS selectors.

vmuzikar commented on testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/clientscopes/ClientScopesEvaluateForm.java:
```diff
@@ -81,6 +82,10 @@
     private WebElement accessTokenTextArea;
 
+    @FindBy(id = "s2id_users")
+    private ClientSelectModal clientsInput;
```
@saibot94 Thanks for the update! Even though it might be a bit confusing that we're using `ClientSelectModal` Page Fragment to select users, I'm gonna approve. ;) Thank you again for your contribution!

Cracky5457 commented on services/src/main/java/org/keycloak/services/resources/admin/ClientsResource.java:
```diff
@@ -91,15 +89,22 @@ public ClientsResource(RealmModel realm, AdminPermissionEvaluator auth, AdminEve
-    public List<ClientRepresentation> getClients(@QueryParam("clientId") String clientId, @QueryParam("viewableOnly") @DefaultValue("false") boolean viewableOnly) {
+    public List<ClientRepresentation> getClients(@QueryParam("clientId") String clientId,
+                                                 @QueryParam("viewableOnly") @DefaultValue("false") boolean viewableOnly,
+                                                 @QueryParam("search") @DefaultValue("false") boolean search,
```
I think you may add a new method in `public interface ClientsResource` with all theses params. I'm pretty sure about that but wait for @stianst  comfirmation. Also for me you should rewrite this function to separate the `findByClientId` and the `findAll` implementation and so your `search` query param become a `String` empty by default to be coherent with other search function in the application ( you can look on the search in `GroupsResource` ), on this wait for @stianst advice too because your solution works but it's not coherent with others endpoints.

saibot94 replied:
I understand the need, it would be nice to have this separation. However, I think that this sort of improvement could come as a separate refactoring task / story, it's not particularly in the scope of the initial development.

I'd be happy to contribute more, once we get this one rolling, as it's been a couple of months since it's been open already.

Cracky5457 replied:
For the second point about separate the function it's not just refactoring, if you define your api with `search` param as boolean and that users start to use it like that, we will have to keep it like that or release an other version different of the api. It sound a bit weird to me. But regarding the current code I don't see an easy solution, this getClientById should probably not have been implemented like that at the beginining because the current params `clientId` is for an `exact` search which can just return one result, should have be a path param probably in a different endpoint !

So I don't know maybe @stianst will have an idea I don't know what happend if we have two implementation using `viewableOnly` on the same path, we will probably have a jaxrs conflict somewhere.

Anyway if you keep things like that, in the `ClientsResource` interface you should add 

```
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public List<ClientRepresentation> searchByClientId(@QueryParam("clientId") String clientId,
                                                 @QueryParam("viewableOnly") @DefaultValue("false") boolean viewableOnly,
                                                 @QueryParam("search") @DefaultValue("false") boolean search,
                                                 @QueryParam("first") Integer firstResult,
                                                 @QueryParam("max") Integer maxResults) ;
```

stianst replied:
The interface ClientsResource in admin-client should have a new method, with the extra param.

For refactoring of the Admin REST APIs - the Admin REST APIs are not extremely nice as they are and there are lots of inconsistencies. What we plan to do when we get time to do it is to introduce new smaller versioned APIs. So there will be a clients API, a users API, etc. each with their own versions.

saibot94 committed:
Address comments on PR about returning null

saibot94 commented:
Hi @stianst @vmuzikar !

What's the status on this one, would you like me to implement the changes suggested by @Cracky5457 ?

vmuzikar commented:
> Hi @stianst @vmuzikar !
> 
> What's the status on this one, would you like me to implement the changes suggested by @Cracky5457 ?

I'm ok with it as is, but this is more of a question for @stianst.

stianst commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmAdapter.java:
```diff
@@ -722,6 +721,21 @@ public ClientModel getClientByClientId(String clientId) {
     }
 
+    @Override
+    public List<ClientModel> searchClientByClientId(String clientId, Integer firstResult, Integer maxResults) {
```
Do we even need searchClientByClientId? It will only return a single instance and there's already getClientByClientId.

stianst replied:
+1 to renaming to searchClientsByClientId

stianst commented:
Re-reviewing it now

stianst reviewed (changes requested):
Need to also update https://github.com/keycloak/keycloak/blob/master/integration/admin-client/src/main/java/org/keycloak/admin/client/resource/ClientsResource.java. With overloaded methods for those that have new parameters, so backwards compatiblity isn't broken.

I'd also like to see tests added for the REST endpoints directly by adding a test to https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/ClientTest.java.

stianst commented:
This is getting very close and I'm pretty much happy with it. Only things missing is the above (update admin-client) and add tests for REST endpoints directly.

@vmuzikar did you get a chance to run the admin console tests locally?

vmuzikar commented:
> @vmuzikar did you get a chance to run the admin console tests locally?

Yes, I ran them when I originally approved the PR. I guess there were no UI changes since then, but I can re-run them again.

stianst commented:
@vmuzikar we can wait with re-running them when we've approved the PR. From my POV the only thing missing is admin-client and some tests for REST APIs directly by extending the ClientTest.

stianst commented:
Completed this work in the following PR: https://github.com/keycloak/keycloak/pull/6567

saibot94 commented:
Thanks a lot for picking it up! 

Really sorry for not having the time to finish up with the `admin-client` endpoint and client tests (although it should've been rather quick and trivial).

stianst commented:
@saibot94 np. You did all the hard work :)
